### ISSUE TITLE
Debugging C with Haskell's Divisible

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#4227]
Debugging C with Haskell's Divisible<br>
<br>
http://bit.ly/2wWTcYl<br>
<br>
About 0 Minutes<br>
<br>
<div>
<div> <p>A good type system covers a lot of the small bugs. Unit tests, careful design, and a sound mind can get you through some bigger ones. But sometimes complex bugs in large codebases call for heavy-duty debugging tools.</p> <p>One such tool is <strong>Delta Debugging</strong>, which repeatedly shrinks recursive data structures to find a minimal counterexample that still exhibits the bug. You’ll appreciate the technique if you’ve ever used <code>git bisect</code> to locate a small breaking change in a large codebase.</p> <p>This article:</p> <ul>
<li>Introduces the Delta Debugging technique</li> <li>Implements a general-purpose delta-debugging tool</li> <li>Uses Haskell’s FFI to control a C chess engine</li> <li>Locates an error introduced into the chess engine</li> </ul>
<h2 id="the-setup">The Setup</h2> <p>In <a href="http://www.michaelburge.us/2017/09/10/injecting-shellcode-to-speed-up-amazon-redshift.html">my last article</a>, I implemented a chess engine in C. I’ve introduced an error into its move generator. Recall that the move generation involved 5 functions:</p> <figure><pre><code><span>typedef</span> <span>...</span> <span>gamestate</span><span>;</span> <span>typedef</span> <span>...</span> <span>iterator</span><span>;</span> <span>typedef</span> <span>...</span> <span>move</span><span>;</span> <span>// Move generation </span><span>iterator</span> <span>mkIterator</span><span>(</span><span>gamestate</span> <span>g</span><span>);</span> <span>iterator</span> <span>advance_iterator</span><span>(</span><span>gamestate</span> <span>g</span><span>,</span> <span>iterator</span> <span>i</span><span>);</span> <span>gamestate</span> <span>apply_move</span><span>(</span><span>gamestate</span> <span>g</span><span>,</span> <span>move</span> <span>m</span><span>);</span> <span>bool</span> <span>is_iterator_finished</span><span>(</span><span>iterator</span> <span>x</span><span>);</span> <span>move</span> <span>dereference_iterator</span><span>(</span><span>iterator</span> <span>i</span><span>);</span> <span>void</span> <span>print_fen</span><span>(</span><span>gamestate</span> <span>g</span><span>,</span> <span>char</span> <span>*</span><span>buffer</span><span>);</span> <span>gamestate</span> <span>new_game</span><span>();</span> <span>// The actual function used to define the bug </span><span>uint64_t</span> <span>perft</span><span>(</span><span>gamestate</span> <span>g</span><span>,</span> <span>int</span> <span>depth</span><span>);</span></code></pre></figure><p>The “gold standard” for testing a chess engine is to compare the output of the above <code>perft</code> function with <a href="https://chessprogramming.wikispaces.com/Perft+Results">published reference values</a>. The bug that we’ll be finding is easily visible from the starting state of the board, so <code>new_game</code> will be the only <code>gamestate</code> value I’ll expose from the engine.</p> <p>In the FFI section, I’ll define Haskell wrappers for everything:</p> <figure><pre><code><span>data</span> <span>Gamestate</span> <span>data</span> <span>Iterator</span> <span>data</span> <span>Move</span> <span>newGame</span> <span>::</span> <span>Gamestate</span> <span>printFen</span> <span>::</span> <span>Gamestate</span> <span>-&gt;</span> <span>String</span> <span>mkIterator</span> <span>::</span> <span>Gamestate</span> <span>-&gt;</span> <span>Iterator</span> <span>advanceIterator</span> <span>::</span> <span>Gamestate</span> <span>-&gt;</span> <span>Iterator</span> <span>-&gt;</span> <span>Iterator</span> <span>perft</span> <span>::</span> <span>Gamestate</span> <span>-&gt;</span> <span>Int32</span> <span>-&gt;</span> <span>Word64</span> <span>applyMove</span> <span>::</span> <span>Gamestate</span> <span>-&gt;</span> <span>Move</span> <span>-&gt;</span> <span>Gamestate</span> <span>isIteratorFinished</span> <span>::</span> <span>Iterator</span> <span>-&gt;</span> <span>Bool</span> <span>dereferenceIterator</span> <span>::</span> <span>Iterator</span> <span>-&gt;</span> <span>Move</span></code></pre></figure><p>To detect the bug, we’ll need a reference version of <code>perft</code> to compare ours against. I’ll use <a href="http://www.rocechess.ch/perft.html">Roce Chess</a>, so our Haskell will literally invoke a <code>roce38</code> process, set the board state, and ask it to calculate a <code>perft</code> value.</p> <figure><pre><code><span>type</span> <span>Move'</span> <span>=</span> <span>String</span> <span>reference_perft</span> <span>::</span> <span>Gamestate</span> <span>-&gt;</span> <span>Int32</span> <span>-&gt;</span> <span>Word64</span> <span>reference_moves</span> <span>::</span> <span>Gamestate</span> <span>-&gt;</span> <span>[</span><span>Move'</span><span>]</span></code></pre></figure><p>I’ve removed the <code>IO</code> constraint using <code>unsafePerformIO</code> to simplify the article’s progression, but the final solution will be general enough to work over <code>IO</code>.</p> <h2 id="reducible">Reducible</h2> <p>Here’s a variation on <a href="https://hackage.haskell.org/package/QuickCheck-2.10.0.1/docs/Test-QuickCheck-Arbitrary.html">QuickCheck’s</a> <code>Arbitrary</code> typeclass:</p> <figure><pre><code><span>-- | Gives all of the single-step reductions for a type.</span> <span>-- | It is recommended to put the largest reductions first.</span> <span>class</span> <span>Reducible</span> <span>a</span> <span>where</span> <span>reductions</span> <span>::</span> <span>a</span> <span>-&gt;</span> <span>[</span><span>a</span><span>]</span> <span>minimize</span> <span>::</span> <span>Reducible</span> <span>a</span> <span>=&gt;</span> <span>(</span><span>a</span> <span>-&gt;</span> <span>Bool</span><span>)</span> <span>-&gt;</span> <span>a</span> <span>-&gt;</span> <span>a</span> <span>minimize</span> <span>f</span> <span>x</span> <span>=</span> <span>case</span> <span>find</span> <span>f</span> <span>$</span> <span>reductions</span> <span>x</span> <span>of</span> <span>Nothing</span> <span>-&gt;</span> <span>x</span> <span>Just</span> <span>y</span> <span>-&gt;</span> <span>minimize</span> <span>f</span> <span>y</span></code></pre></figure><p>If you have an HTML file that crashes your web browser, <code>minimize</code> would remove tags or characters until it is a minimal crash-causing file. This variant would remove only a single tag or character at a time, while <code>Arbitrary</code> removes entire batches for performance.</p> <p>Recall that <code>perft</code> takes a <code>Gamestate</code> and a <code>Depth</code>. The “smallest possible” means that the <code>Depth</code> is the smallest.</p> <figure><pre><code><span>instance</span> <span>Reducible</span> <span>Gamestate</span> <span>where</span> <span>reductions</span> <span>::</span> <span>Gamestate</span> <span>-&gt;</span> <span>[</span><span>Gamestate</span><span>]</span> <span>reductions</span> <span>g</span> <span>=</span> <span>let</span> <span>loop</span> <span>i</span> <span>=</span> <span>if</span> <span>isIteratorFinished</span> <span>i</span> <span>then</span> <span>[]</span> <span>else</span> <span>let</span> <span>m</span> <span>=</span> <span>dereferenceIterator</span> <span>i</span> <span>g'</span> <span>=</span> <span>applyMove</span> <span>g</span> <span>m</span> <span>i'</span> <span>=</span> <span>advanceIterator</span> <span>g</span> <span>i</span> <span>in</span> <span>g'</span> <span>:</span> <span>loop</span> <span>i'</span> <span>in</span> <span>loop</span> <span>$</span> <span>mkIterator</span> <span>g</span> <span>type</span> <span>Depth</span> <span>=</span> <span>Int32</span> <span>newtype</span> <span>PerftTest</span> <span>=</span> <span>PerftTest</span> <span>(</span><span>Gamestate</span><span>,</span> <span>Depth</span><span>)</span> <span>deriving</span> <span>(</span><span>Show</span><span>)</span> <span>instance</span> <span>Reducible</span> <span>PerftTest</span> <span>where</span> <span>reductions</span> <span>::</span> <span>PerftTest</span> <span>-&gt;</span> <span>[</span><span>PerftTest</span><span>]</span> <span>reductions</span> <span>(</span><span>PerftTest</span> <span>(</span><span>g</span><span>,</span> <span>depth</span><span>))</span> <span>=</span> <span>if</span> <span>depth</span> <span>==</span> <span>1</span> <span>then</span> <span>[]</span> <span>else</span> <span>map</span> <span>(</span><span>\</span><span>g'</span> <span>-&gt;</span> <span>PerftTest</span> <span>(</span><span>g'</span><span>,</span> <span>depth</span><span>-</span><span>1</span><span>))</span> <span>$</span> <span>reductions</span> <span>g</span></code></pre></figure><p>And here’s how to apply this to find a specific position and move that differ between the two engines:</p> <figure><pre><code><span>module</span> <span>Main</span> <span>where</span> <span>import</span> <span>Chess</span> <span>import</span> <span>Reducible</span> <span>checkBug</span> <span>::</span> <span>PerftTest</span> <span>-&gt;</span> <span>Bool</span> <span>checkBug</span> <span>(</span><span>PerftTest</span> <span>(</span><span>g</span><span>,</span> <span>depth</span><span>))</span> <span>=</span> <span>perft</span> <span>g</span> <span>depth</span> <span>/=</span> <span>reference_perft</span> <span>g</span> <span>depth</span> <span>compareMoves</span> <span>::</span> <span>Gamestate</span> <span>-&gt;</span> <span>([</span><span>Move'</span><span>],</span> <span>[</span><span>Move'</span><span>])</span> <span>compareMoves</span> <span>g</span> <span>=</span> <span>let</span> <span>lefts</span> <span>=</span> <span>map</span> <span>printMove</span> <span>$</span> <span>moves</span> <span>g</span> <span>rights</span> <span>=</span> <span>reference_moves</span> <span>g</span> <span>in</span> <span>(</span><span>lefts</span> <span>\\</span> <span>rights</span><span>,</span> <span>rights</span> <span>\\</span> <span>lefts</span><span>)</span> <span>instance</span> <span>Show</span> <span>Gamestate</span> <span>where</span> <span>show</span> <span>g</span> <span>=</span> <span>printFen</span> <span>g</span> <span>main</span> <span>::</span> <span>IO</span> <span>()</span> <span>main</span> <span>=</span> <span>do</span> <span>let</span> <span>g</span> <span>=</span> <span>newGame</span> <span>i</span> <span>=</span> <span>mkIterator</span> <span>g</span> <span>let</span> <span>(</span><span>PerftTest</span> <span>(</span><span>g'</span><span>,</span> <span>depth</span><span>))</span> <span>=</span> <span>minimize</span> <span>checkBug</span> <span>$</span> <span>PerftTest</span> <span>(</span><span>g</span><span>,</span> <span>3</span><span>)</span> <span>putStrLn</span> <span>$</span> <span>show</span> <span>$</span> <span>g'</span> <span>putStrLn</span> <span>$</span> <span>show</span> <span>$</span> <span>compareMoves</span> <span>g'</span> <span>{- Prints: r1bqkbnr/pppppppp/n7/8/8/4P3/PPPP1PPP/RNBQKBNR w KQkq - 0 1 ([],["e1e2"]) -}</span></code></pre></figure><p>The FEN string can be pasted directly into a program like <code>XBoard</code> to visualize the position, and “e1e2” is a move that Rocechess emits that my engine doesn’t.</p> <p>At this point, you would have to dig into the C to see why the King at e1 is not being allowed to move northward to e2. I’ll spoil that by revealing the change that I made:</p> <figure><pre><code><span>private</span> <span>uint64_t</span> <span>valid_king_moves</span><span>(</span><span>gamestate</span> <span>g</span><span>,</span> <span>int</span> <span>idx</span><span>)</span> <span>{</span> <span>uint64_t</span> <span>ret</span> <span>=</span> <span>bit</span><span>(</span><span>move_direction</span><span>(</span><span>idx</span><span>,</span> <span>DIRECTION_EAST</span><span>))</span> <span>|</span> <span>bit</span><span>(</span><span>move_direction</span><span>(</span><span>idx</span><span>,</span> <span>DIRECTION_WEST</span><span>))</span> <span>|</span> <span>// bit(move_direction(idx, DIRECTION_NORTH)) | </span> <span>bit</span><span>(</span><span>move_direction</span><span>(</span><span>idx</span><span>,</span> <span>DIRECTION_SOUTH</span><span>))</span> <span>|</span> <span>bit</span><span>(</span><span>move_direction</span><span>(</span><span>idx</span><span>,</span> <span>DIRECTION_NORTHEAST</span><span>))</span> <span>|</span> <span>bit</span><span>(</span><span>move_direction</span><span>(</span><span>idx</span><span>,</span> <span>DIRECTION_SOUTHEAST</span><span>))</span> <span>|</span> <span>bit</span><span>(</span><span>move_direction</span><span>(</span><span>idx</span><span>,</span> <span>DIRECTION_NORTHWEST</span><span>))</span> <span>|</span> <span>bit</span><span>(</span><span>move_direction</span><span>(</span><span>idx</span><span>,</span> <span>DIRECTION_SOUTHWEST</span><span>))</span> <span>;</span> <span>...</span> <span>return</span> <span>ret</span><span>;</span> <span>}</span></code></pre></figure><p>My experience across many different software projects is that most of the work in debugging is in creating a small isolated test case that demonstrates the problem, and not in the actual fix.</p> <h2 id="divisible">Divisible</h2> <p><a href="https://hackage.haskell.org/package/contravariant-1.4/docs/Data-Functor-Contravariant-Divisible.html"><code>Divisible</code></a> is the main reason I wrote this article, and allows us to generalize <code>minimize</code> above. Since the context is debugging a chess engine, I’ll use the language of deductive logic to describe how we’ll use it.</p> <p>The idea is that to prove that the value of <code>perft g 6</code> is correct, you need two components:</p> <ul>
<li>A way to split <code>perft g 6</code> into subproblems <code>perft g' 5</code>.</li> <li>A way to prove base cases correct: <code>perft g 0 == 1</code>
</li> </ul>
<p>Since we’re debugging, such a proof is impossible because we have a failing testcase in-hand. However, we get to define what “proof” means in our context, so we’ll use “Correct, or produces a concrete minimal counterexample”:</p> <figure><pre><code><span>-- A testcase a can be decomposed into smaller child problems.</span> <span>-- A Counterexample finder scans an a and produces:</span> <span>-- 1. Nothing: A proof that all children are correct</span> <span>-- 2. Just b: Evidence of type b that some child is erroneous</span> <span>newtype</span> <span>Cx</span> <span>b</span> <span>a</span> <span>=</span> <span>Cx</span> <span>{</span> <span>unCx</span> <span>::</span> <span>a</span> <span>-&gt;</span> <span>Maybe</span> <span>b</span> <span>}</span></code></pre></figure><p>There are actually 3 different typeclasses, corresponding to 3 different proof techniques:</p> <h3 id="contravariant">Contravariant</h3> <p>In most programs, you start with values obtained from sensory data like files or system calls, and apply functions to convert those into more abstract or useful information. This is reversed when debugging: We start with a goal in mind(a failing testcase), and work “backwards” to find the first place with a problem. Thus, <code>Contravariant</code>, <code>Divisible</code>, and <code>Decidable</code> have type signatures that seem “impossible” in forward-thinking programs:</p> <figure><pre><code><span>class</span> <span>Contravariant</span> <span>f</span> <span>where</span> <span>contramap</span> <span>::</span> <span>(</span><span>a</span> <span>-&gt;</span> <span>b</span><span>)</span> <span>-&gt;</span> <span>f</span> <span>b</span> <span>-&gt;</span> <span>f</span> <span>a</span></code></pre></figure><p>In a proof context, <code>contramap</code> is a way of reducing a goal of type <code>a</code> to a goal of type <code>b</code>: If <code>f b</code> represents a sensor that detects erroneous <code>b</code> values, then we can convert it into a sensor that detects erroneous <code>a</code> values by converting the <code>a</code>s to <code>b</code>s and then scanning the <code>b</code>s.</p> <p>A goal will usually have a function type like <code>type Goal a = a -&gt; Bool</code>. The mystery behind <code>Contravariant</code>’s impossible signature goes away when you realize we don’t have a value of type <code>a</code> or <code>b</code> in the first place: We only have a function, so we’re waiting on code upstream to deliver one to us.</p> <p>Here’s a concrete implementation</p> <figure><pre><code><span>instance</span> <span>Contravariant</span> <span>(</span><span>Cx</span> <span>b</span><span>)</span> <span>where</span> <span>contramap</span> <span>f</span> <span>(</span><span>Cx</span> <span>g</span><span>)</span> <span>=</span> <span>Cx</span> <span>(</span><span>g</span> <span>.</span> <span>f</span><span>)</span></code></pre></figure><h3 id="divisible-1">Divisible</h3> <p><code>contramap</code> lets us solve a goal by relating it to an easier goal, and then solving the easier one. We might also want to split the goal into multiple easier subgoals:</p> <figure><pre><code><span>class</span> <span>Contravariant</span> <span>f</span> <span>=&gt;</span> <span>Divisible</span> <span>f</span> <span>where</span> <span>divide</span> <span>::</span> <span>(</span><span>a</span> <span>-&gt;</span> <span>(</span><span>b</span><span>,</span> <span>c</span><span>))</span> <span>-&gt;</span> <span>f</span> <span>b</span> <span>-&gt;</span> <span>f</span> <span>c</span> <span>-&gt;</span> <span>f</span> <span>a</span> <span>conquer</span> <span>::</span> <span>f</span> <span>a</span></code></pre></figure><p><code>divide</code> is reasonable enough, but in a proof context <code>conquer</code> seems to let us prove anything at all. <code>divide</code> can be generalized to arbitrary tuples, and <code>conquer</code> is the empty conjuction:</p> <figure><pre><code><span>divide4</span> <span>::</span> <span>Divisible</span> <span>f</span> <span>=&gt;</span> <span>(</span><span>a</span> <span>-&gt;</span> <span>(</span><span>b</span><span>,</span><span>c</span><span>,</span><span>d</span><span>,</span><span>e</span><span>))</span> <span>-&gt;</span> <span>f</span> <span>b</span> <span>-&gt;</span> <span>f</span> <span>c</span> <span>-&gt;</span> <span>f</span> <span>d</span> <span>-&gt;</span> <span>f</span> <span>e</span> <span>-&gt;</span> <span>f</span> <span>a</span> <span>divide3</span> <span>::</span> <span>Divisible</span> <span>f</span> <span>=&gt;</span> <span>(</span><span>a</span> <span>-&gt;</span> <span>(</span><span>b</span><span>,</span><span>c</span><span>,</span><span>d</span><span>))</span> <span>-&gt;</span> <span>f</span> <span>b</span> <span>-&gt;</span> <span>f</span> <span>c</span> <span>-&gt;</span> <span>f</span> <span>d</span> <span>-&gt;</span> <span>f</span> <span>a</span> <span>divide2</span> <span>::</span> <span>Divisible</span> <span>f</span> <span>=&gt;</span> <span>(</span><span>a</span> <span>-&gt;</span> <span>(</span><span>b</span><span>,</span><span>c</span><span>))</span> <span>-&gt;</span> <span>f</span> <span>b</span> <span>-&gt;</span> <span>f</span> <span>c</span> <span>-&gt;</span> <span>f</span> <span>a</span> <span>divide1</span> <span>::</span> <span>Contravariant</span> <span>f</span> <span>=&gt;</span> <span>(</span><span>a</span> <span>-&gt;</span> <span>(</span><span>b</span><span>))</span> <span>-&gt;</span> <span>f</span> <span>b</span> <span>-&gt;</span> <span>f</span> <span>a</span> <span>divide0</span> <span>::</span> <span>Divisible</span> <span>f</span> <span>=&gt;</span> <span>(</span><span>a</span> <span>-&gt;</span> <span>()</span><span>)</span> <span>-&gt;</span> <span>f</span> <span>a</span> <span>divide0</span> <span>_</span> <span>=</span> <span>conquer</span> <span>divide1</span> <span>=</span> <span>contramap</span> <span>divide2</span> <span>=</span> <span>divide</span> <span>divide3</span> <span>split3</span> <span>pb</span> <span>pc</span> <span>pd</span> <span>=</span> <span>divide</span> <span>(</span><span>reassociate</span> <span>.</span> <span>split3</span><span>)</span> <span>pb</span> <span>$</span> <span>divide</span> <span>id</span> <span>pc</span> <span>pd</span> <span>where</span> <span>reassociate</span> <span>(</span><span>x</span><span>,</span><span>y</span><span>,</span><span>z</span><span>)</span> <span>=</span> <span>(</span><span>x</span><span>,(</span><span>y</span><span>,</span><span>z</span><span>))</span> <span>divide4</span> <span>split4</span> <span>pb</span> <span>pc</span> <span>pd</span> <span>pe</span> <span>=</span> <span>divide</span> <span>(</span><span>reassociate4</span> <span>.</span> <span>split4</span><span>)</span> <span>pb</span> <span>$</span> <span>divide3</span> <span>id</span> <span>pc</span> <span>pd</span> <span>pe</span> <span>where</span> <span>reassociate4</span> <span>(</span><span>x</span><span>,</span><span>y</span><span>,</span><span>z</span><span>,</span><span>w</span><span>)</span> <span>=</span> <span>(</span><span>x</span><span>,(</span><span>y</span><span>,</span><span>z</span><span>,</span><span>w</span><span>))</span></code></pre></figure><p>In general, <code>divide</code> takes a collection of subproblems and solutions for those subproblems, and combines them into a solution for the larger problem. Here’s the tuple example generalized to arbitrarily-large lists:</p> <figure><pre><code><span>divideList</span> <span>::</span> <span>Divisible</span> <span>f</span> <span>=&gt;</span> <span>[(</span><span>a</span> <span>-&gt;</span> <span>b</span><span>,</span> <span>f</span> <span>b</span><span>)]</span> <span>-&gt;</span> <span>f</span> <span>a</span> <span>divideList</span> <span>=</span> <span>\</span><span>case</span> <span>[]</span> <span>-&gt;</span> <span>conquer</span> <span>((</span><span>f</span><span>,</span> <span>px</span><span>)</span><span>:</span><span>xs</span><span>)</span> <span>-&gt;</span> <span>divide</span> <span>(</span><span>reproject</span> <span>f</span><span>)</span> <span>px</span> <span>(</span><span>divideList</span> <span>xs</span><span>)</span> <span>where</span> <span>reproject</span> <span>::</span> <span>(</span><span>a</span> <span>-&gt;</span> <span>b</span><span>)</span> <span>-&gt;</span> <span>(</span><span>a</span> <span>-&gt;</span> <span>(</span><span>b</span><span>,</span><span>a</span><span>))</span> <span>reproject</span> <span>f</span> <span>=</span> <span>\</span><span>x</span> <span>-&gt;</span> <span>(</span><span>f</span> <span>x</span><span>,</span> <span>x</span><span>)</span></code></pre></figure><p>The <code>divideList</code> above isn’t fully-general because it requires every subproblem to be of the same type.</p> <p>Then, <code>conquer</code> is an assertion that no further subproblems need to be checked. Recall that our <code>Cx</code> type means “Either all children are correct, or we’ve found suspicious evidence”. <code>conquer</code> will correspond to this first case.</p> <p>Here’s how to implement <code>divide</code> for our counterexample detector:</p> <figure><pre><code><span>instance</span> <span>Divisible</span> <span>(</span><span>Cx</span> <span>b</span><span>)</span> <span>where</span> <span>divide</span> <span>split</span> <span>(</span><span>Cx</span> <span>left</span><span>)</span> <span>(</span><span>Cx</span> <span>right</span><span>)</span> <span>=</span> <span>Cx</span> <span>$</span> <span>\</span><span>x</span> <span>-&gt;</span> <span>let</span> <span>(</span><span>lProblem</span><span>,</span> <span>rProblem</span><span>)</span> <span>=</span> <span>split</span> <span>x</span> <span>-- One branch must produce a counterexample</span> <span>in</span> <span>case</span> <span>(</span><span>left</span> <span>lProblem</span><span>,</span> <span>right</span> <span>rProblem</span><span>)</span> <span>of</span> <span>(</span><span>Just</span> <span>x</span><span>,</span> <span>_</span><span>)</span> <span>-&gt;</span> <span>Just</span> <span>x</span> <span>(</span><span>_</span><span>,</span> <span>Just</span> <span>x</span><span>)</span> <span>-&gt;</span> <span>Just</span> <span>x</span> <span>(</span><span>Nothing</span><span>,</span> <span>Nothing</span><span>)</span> <span>-&gt;</span> <span>Nothing</span> <span>conquer</span> <span>=</span> <span>Cx</span> <span>$</span> <span>const</span> <span>Nothing</span></code></pre></figure><p>Since both subproblems must be clean in order for the parent to be clean, we can can return early if we find an error in the left branch.</p> <h3 id="decidable">Decidable</h3> <p><code>contramap</code> corresponds to the logical <a href="https://en.wikipedia.org/wiki/Proof_by_contrapositive">contrapositive</a>, while <code>divide</code> corresponds to conjuction. The other major tool is disjunction, which is provided by <code>choose</code>.</p> <figure><pre><code><span>class</span> <span>Divisible</span> <span>f</span> <span>=&gt;</span> <span>Decidable</span> <span>f</span> <span>where</span> <span>lose</span> <span>::</span> <span>(</span><span>a</span> <span>-&gt;</span> <span>Void</span><span>)</span> <span>-&gt;</span> <span>f</span> <span>a</span> <span>choose</span> <span>::</span> <span>(</span><span>a</span> <span>-&gt;</span> <span>Either</span> <span>b</span> <span>c</span><span>)</span> <span>-&gt;</span> <span>f</span> <span>b</span> <span>-&gt;</span> <span>f</span> <span>c</span> <span>-&gt;</span> <span>f</span> <span>a</span></code></pre></figure><p><code>lose</code> is a placeholder when you know that a branch in <code>choose</code> is impossible yet are required to provide a proof. There are only 3 possible implementations:</p> <ul>
<li>An exception, using <code>undefined</code> or <a href="https://hackage.haskell.org/package/void-0.7.1/docs/Data-Void.html"><code>absurd</code></a>.</li> <li><code>const conquer</code></li> <li>A constant distinct from <code>conquer</code>
</li> </ul>
<p>I don’t see any examples of the 3rd type in the <code>contravariant</code> package, but I imagine that it could be used to write backwards backtracking algorithms similarly to how <a href="http://hackage.haskell.org/package/base-4.10.0.0/docs/Control-Monad.html#v:guard"><code>guard :: Bool -&gt; [()]</code></a> is used to <a href="https://blog.jle.im/entry/practical-fun-with-monads-introducing-monadplus.html">prune branches</a> using <code>MonadPlus</code>.</p> <p><code>choose</code> in our context says that if you are trying to locate errors in a value of type <code>a</code>, can split it into one of two subproblems, and can locate errors in both subproblems, then you can locate errors by figuring out which subproblem is appropriate for <code>a</code> and reusing its proof. Here’s the implementation:</p> <figure><pre><code><span>instance</span> <span>Decidable</span> <span>(</span><span>Cx</span> <span>b</span><span>)</span> <span>where</span> <span>lose</span> <span>_</span> <span>=</span> <span>conquer</span> <span>choose</span> <span>split</span> <span>left</span> <span>right</span> <span>=</span> <span>Cx</span> <span>$</span> <span>\</span><span>x</span> <span>-&gt;</span> <span>-- Both branches must produce a counterexample</span> <span>case</span> <span>split</span> <span>x</span> <span>of</span> <span>Left</span> <span>l</span> <span>-&gt;</span> <span>unCx</span> <span>left</span> <span>l</span> <span>Right</span> <span>r</span> <span>-&gt;</span> <span>unCx</span> <span>right</span> <span>r</span></code></pre></figure><p>The main difference between <code>Decidable</code> and <code>Divisible</code> is that <code>Decidable</code> branches to precisely one of its subproblems, while <code>Divisible</code> combines subproblems together.</p> <p>I’ll implement <code>decideList</code> to give a feel for how <code>Decidable</code> works in practice before moving on to reimplement <code>minimize</code>.</p> <figure><pre><code><span>decideList</span> <span>::</span> <span>forall</span> <span>a</span> <span>b</span> <span>f</span><span>.</span> <span>Decidable</span> <span>f</span> <span>=&gt;</span> <span>(</span><span>a</span> <span>-&gt;</span> <span>(</span><span>Integer</span><span>,</span> <span>b</span><span>))</span> <span>-&gt;</span> <span>[</span><span>f</span> <span>b</span><span>]</span> <span>-&gt;</span> <span>f</span> <span>a</span> <span>decideList</span> <span>f</span> <span>(</span><span>xp</span><span>:</span><span>xps</span><span>)</span> <span>=</span> <span>choose</span> <span>caseConstructor</span> <span>xp</span> <span>(</span><span>decideList</span> <span>nextF</span> <span>xps</span><span>)</span> <span>where</span> <span>caseConstructor</span> <span>::</span> <span>a</span> <span>-&gt;</span> <span>Either</span> <span>b</span> <span>a</span> <span>caseConstructor</span> <span>x</span> <span>=</span> <span>let</span> <span>(</span><span>c</span><span>,</span> <span>w</span><span>)</span> <span>=</span> <span>f</span> <span>x</span> <span>in</span> <span>case</span> <span>c</span> <span>of</span> <span>0</span> <span>-&gt;</span> <span>Left</span> <span>w</span> <span>n</span> <span>-&gt;</span> <span>Right</span> <span>x</span> <span>nextF</span> <span>::</span> <span>a</span> <span>-&gt;</span> <span>(</span><span>Integer</span><span>,</span> <span>b</span><span>)</span> <span>nextF</span> <span>x</span> <span>=</span> <span>let</span> <span>(</span><span>c</span><span>,</span> <span>w</span><span>)</span> <span>=</span> <span>f</span> <span>x</span> <span>in</span> <span>(</span><span>c</span><span>-</span><span>1</span><span>,</span> <span>w</span><span>)</span></code></pre></figure><p>The idea here is that <code>(Integer, b)</code> consists of a branch selector and a value <code>b</code>. For <code>Either</code>, <code>Left</code> would be branch 0, while <code>Right</code> would be branch 1. <code>caseConstructor</code> either chooses the current branch, or threads the <code>a</code> value along to the next one.</p> <p>As with <code>divideList</code>, <code>decideList</code> isn’t fully-general because it requires every subproblem to be of the same type.</p> <h3 id="minimize">minimize</h3> <p>We’re ready to generalize our original <code>minimize</code> function. Here it is again:</p> <figure><pre><code><span>minimize</span> <span>::</span> <span>Reducible</span> <span>a</span> <span>=&gt;</span> <span>(</span><span>a</span> <span>-&gt;</span> <span>Bool</span><span>)</span> <span>-&gt;</span> <span>a</span> <span>-&gt;</span> <span>a</span> <span>minimize</span> <span>f</span> <span>x</span> <span>=</span> <span>case</span> <span>find</span> <span>f</span> <span>$</span> <span>reductions</span> <span>x</span> <span>of</span> <span>Nothing</span> <span>-&gt;</span> <span>x</span> <span>Just</span> <span>y</span> <span>-&gt;</span> <span>minimize</span> <span>f</span> <span>y</span></code></pre></figure><p>We generalize it in 3 steps:</p> <ul>
<li> <p>First, our <code>Cx</code> type contains an <code>a -&gt; Maybe a</code>, which can represent both the <code>a -&gt; Bool</code> and <code>a -&gt; a</code> portions of <code>minimize</code>. So <code>minimize</code> becomes <code>Reducible a =&gt; Cx a a -&gt; Cx a a</code>.</p> </li> <li> <p>Second, I’ve generalized <code>Cx</code> to a <code>Decidable f</code> constraint, causing its signature to be <code>(Decidable f, Reducible a) =&gt; f a -&gt; f a</code>.</p> </li> <li> <p>Third, <code>minimize</code> calls <a href="https://hackage.haskell.org/package/base-4.10.0.0/docs/Data-List.html"><code>find :: (a -&gt; Bool) -&gt; [a] -&gt; Maybe a</code></a>. So I’ve generalized it to <code>findD :: Decidable f =&gt; f a -&gt; f [a]</code>.</p> </li> </ul>
<figure><pre><code><span>import</span> <span>Data.List</span> <span>(</span><span>uncons</span><span>)</span> <span>import</span> <span>Control.Arrow</span> <span>((</span><span>&amp;&amp;&amp;</span><span>))</span> <span>minimizeD</span> <span>::</span> <span>(</span><span>Decidable</span> <span>f</span><span>,</span> <span>Reducible</span> <span>a</span><span>)</span> <span>=&gt;</span> <span>f</span> <span>a</span> <span>-&gt;</span> <span>f</span> <span>a</span> <span>minimizeD</span> <span>pred</span> <span>=</span> <span>divide</span> <span>(</span><span>reductions</span> <span>&amp;&amp;&amp;</span> <span>id</span><span>)</span> <span>(</span><span>findD</span> <span>$</span> <span>minimizeD</span> <span>pred</span><span>)</span> <span>pred</span> <span>findD</span> <span>::</span> <span>Decidable</span> <span>f</span> <span>=&gt;</span> <span>f</span> <span>a</span> <span>-&gt;</span> <span>f</span> <span>[</span><span>a</span><span>]</span> <span>findD</span> <span>p</span> <span>=</span> <span>chooseMaybe</span> <span>uncons</span> <span>conquer</span> <span>(</span><span>divided</span> <span>p</span> <span>(</span><span>findD</span> <span>p</span><span>))</span> <span>chooseMaybe</span> <span>::</span> <span>Decidable</span> <span>f</span> <span>=&gt;</span> <span>(</span><span>a</span> <span>-&gt;</span> <span>Maybe</span> <span>b</span><span>)</span> <span>-&gt;</span> <span>f</span> <span>()</span> <span>-&gt;</span> <span>f</span> <span>b</span> <span>-&gt;</span> <span>f</span> <span>a</span> <span>chooseMaybe</span> <span>p</span> <span>nothing</span> <span>just</span> <span>=</span> <span>choose</span> <span>(</span><span>maybe</span> <span>(</span><span>Left</span> <span>()</span><span>)</span> <span>Right</span> <span>.</span> <span>p</span><span>)</span> <span>nothing</span> <span>just</span> <span>uncons</span> <span>::</span> <span>[</span><span>a</span><span>]</span> <span>-&gt;</span> <span>Maybe</span> <span>(</span><span>a</span><span>,</span> <span>[</span><span>a</span><span>])</span> <span>(</span><span>&amp;&amp;&amp;</span><span>)</span> <span>::</span> <span>(</span><span>a</span> <span>-&gt;</span> <span>b</span><span>)</span> <span>-&gt;</span> <span>(</span><span>a</span> <span>-&gt;</span> <span>c</span><span>)</span> <span>-&gt;</span> <span>(</span><span>a</span> <span>-&gt;</span> <span>(</span><span>b</span><span>,</span><span>c</span><span>))</span></code></pre></figure><p>In our specific context, here’s how you might read these in English:</p> <ul>
<li>
<code>minimizeD</code>: A minimal error is either in one of the <code>reductions</code> from the current node, or the current node itself(<code>id</code>).</li> <li>
<code>findD</code>: A minimal error is either nonexistent because the list was empty(<code>conquer</code>), the current element(<code>p</code>), or in the remainder of the list(<code>findD p</code>).</li> </ul>
<p>And here’s a function with a simpler type that can be easily applied:</p> <figure><pre><code><span>minimizeD''</span> <span>::</span> <span>PerftTest</span> <span>-&gt;</span> <span>Maybe</span> <span>PerftTest</span> <span>minimizeD''</span> <span>x</span> <span>=</span> <span>unCx</span> <span>(</span><span>minimizeD</span> <span>(</span><span>Cx</span> <span>$</span> <span>liftPred</span> <span>checkBug</span><span>))</span> <span>x</span> <span>where</span> <span>liftPred</span> <span>::</span> <span>(</span><span>a</span> <span>-&gt;</span> <span>Bool</span><span>)</span> <span>-&gt;</span> <span>(</span><span>a</span> <span>-&gt;</span> <span>Maybe</span> <span>a</span><span>)</span> <span>liftPred</span> <span>pred</span> <span>=</span> <span>\</span><span>x</span> <span>-&gt;</span> <span>if</span> <span>pred</span> <span>x</span> <span>then</span> <span>Just</span> <span>x</span> <span>else</span> <span>Nothing</span> <span>testCase</span> <span>::</span> <span>Int</span> <span>-&gt;</span> <span>PerftTest</span> <span>testCase</span> <span>n</span> <span>=</span> <span>PerftTest</span> <span>(</span><span>newGame</span><span>,</span> <span>n</span><span>)</span> <span>exampleInvocation</span> <span>=</span> <span>minimizeD''</span> <span>$</span> <span>testCase</span> <span>3</span></code></pre></figure><h3 id="performance">Performance</h3> <p>One of the reasons I wrote an FFI to a C chess engine was to demonstrate that the terms and typeclasses I’m pushing around aren’t just abstract nonsense but have actual computational meaning. Nobody criticizes C for being too abstract, so I felt that it would give a good counterbalance to the control code in <code>minimizeD</code>.</p> <p>Recall that <code>pred</code> instantiates a separate system process every time it’s called. It seems reasonable that a good solution should create as few subprocesses as possible. Even ignoring my use of <code>unsafePerformIO</code>, a good solution should avoid recursing too deeply or invoking expensive functions unnecessarily.</p> <p>I’ll add a global counter that we increment every time we create a Rocechess process. I’ll use the counter to detect any hidden performance issues.</p> <figure><pre><code><span>import</span> <span>Data.IORef</span> <span>counter</span> <span>=</span> <span>unsafePerformIO</span> <span>$</span> <span>newIORef</span> <span>0</span> <span>tickCounter</span> <span>::</span> <span>IO</span> <span>()</span> <span>tickCounter</span> <span>=</span> <span>modifyIORef'</span> <span>counter</span> <span>(</span><span>+</span><span>1</span><span>)</span> <span>timeTicks</span> <span>::</span> <span>Show</span> <span>a</span> <span>=&gt;</span> <span>a</span> <span>-&gt;</span> <span>IO</span> <span>Int</span> <span>timeTicks</span> <span>x</span> <span>=</span> <span>do</span> <span>ticks</span> <span>&lt;-</span> <span>readIORef</span> <span>counter</span> <span>putStrLn</span> <span>$</span> <span>show</span> <span>x</span> <span>ticks'</span> <span>&lt;-</span> <span>readIORef</span> <span>counter</span> <span>return</span> <span>$</span> <span>ticks'</span> <span>-</span> <span>ticks</span></code></pre></figure><p><code>timeTicks</code> forces its input by printing it as output, and compares the counter before and after. Here’s the output using the original <code>minimize</code>:</p> <figure><pre><code><span>timeTicks</span> <span>$</span> <span>minimize</span> <span>checkBug</span> <span>$</span> <span>testCase</span> <span>3</span> <span>-- PRINTS</span> <span>-- PerftTest (r1bqkbnr/pppppppp/n7/8/8/4P3/PPPP1PPP/RNBQKBNR w KQkq - 0 1,1)</span> <span>-- 14</span></code></pre></figure><p>So a total of 14 processes are spawned calculating reference <code>perft</code> values to find this minimal testcase. How does our slick one-liner fare?</p> <figure><pre><code><span>timeTicks</span> <span>$</span> <span>minimizeD''</span> <span>$</span> <span>testCase</span> <span>3</span> <span>-- PRINTS</span> <span>-- Just (PerftTest (r1bqkbnr/pppppppp/n7/8/8/4P3/PPPP1PPP/RNBQKBNR w KQkq - 0 1,1))</span> <span>-- 253</span></code></pre></figure><p>The generalized algorithm generates 18x as many calls as our simple one, to find the same failing game position.</p> <h3 id="refinable">Refinable</h3> <p>To solve this problem, I’ll first diagnose the problem and then introduce a new typeclass <code>Refinable</code> to solve it.</p> <p>Let’s compare the two side-by-side:</p> <figure><pre><code><span>minimizeD</span> <span>::</span> <span>(</span><span>Decidable</span> <span>f</span><span>,</span> <span>Reducible</span> <span>a</span><span>)</span> <span>=&gt;</span> <span>f</span> <span>a</span> <span>-&gt;</span> <span>f</span> <span>a</span> <span>minimizeD</span> <span>pred</span> <span>=</span> <span>divide</span> <span>(</span><span>reductions</span> <span>&amp;&amp;&amp;</span> <span>id</span><span>)</span> <span>(</span><span>findD</span> <span>$</span> <span>minimizeD</span> <span>pred</span><span>)</span> <span>pred</span> <span>minimize</span> <span>::</span> <span>Reducible</span> <span>a</span> <span>=&gt;</span> <span>(</span><span>a</span> <span>-&gt;</span> <span>Bool</span><span>)</span> <span>-&gt;</span> <span>a</span> <span>-&gt;</span> <span>a</span> <span>minimize</span> <span>f</span> <span>x</span> <span>=</span> <span>case</span> <span>find</span> <span>f</span> <span>$</span> <span>reductions</span> <span>x</span> <span>of</span> <span>Nothing</span> <span>-&gt;</span> <span>x</span> <span>Just</span> <span>y</span> <span>-&gt;</span> <span>minimize</span> <span>f</span> <span>y</span></code></pre></figure><p>Notice that <code>minimize</code> calls <code>f</code> on each element emitted by <code>reductions</code>, while <code>minimizeD</code> calls <code>minimizeD pred</code>. With <code>minimize</code>, there is no recursion if the node evaluates to <code>Nothing</code>, since we know all children are correct.</p> <p>We could stop the recursion by only passing <code>pred</code> to <code>findD</code>:</p> <figure><pre><code> <span>minimizeD</span> <span>::</span> <span>(</span><span>Decidable</span> <span>f</span><span>,</span> <span>Reducible</span> <span>a</span><span>)</span> <span>=&gt;</span> <span>f</span> <span>a</span> <span>-&gt;</span> <span>f</span> <span>a</span> <span>minimizeD</span> <span>pred</span> <span>=</span> <span>divide</span> <span>(</span><span>reductions</span> <span>&amp;&amp;&amp;</span> <span>id</span><span>)</span> <span>(</span><span>findD</span> <span>$</span> <span>pred</span><span>)</span> <span>pred</span></code></pre></figure><p>But now the counterexample is only 1 step smaller, and not a minimal counterexample.</p> <p>What we need is to <strong>refine</strong> the failure case. First we run <code>pred</code> to scan for an element in a parent node(a <strong>coarse</strong> failure), then we run <code>minimizeD pred</code> to reduce it once we’ve confirmed that one is present(a <strong>fine</strong> failure).</p> <p>Here’s how to implement that:</p> <figure><pre><code><span>class</span> <span>Divisible</span> <span>f</span> <span>=&gt;</span> <span>Refinable</span> <span>f</span> <span>where</span> <span>implies</span> <span>::</span> <span>f</span> <span>a</span> <span>-&gt;</span> <span>f</span> <span>a</span> <span>-&gt;</span> <span>f</span> <span>a</span> <span>instance</span> <span>Refinable</span> <span>(</span><span>Cx</span> <span>b</span><span>)</span> <span>where</span> <span>implies</span> <span>l</span> <span>r</span> <span>=</span> <span>Cx</span> <span>$</span> <span>\</span><span>x</span> <span>-&gt;</span> <span>case</span> <span>unCx</span> <span>l</span> <span>x</span> <span>of</span> <span>Nothing</span> <span>-&gt;</span> <span>Nothing</span> <span>Just</span> <span>_</span> <span>-&gt;</span> <span>unCx</span> <span>r</span> <span>x</span></code></pre></figure><p>Just as <code>divide</code> lets us abort a proof as soon as we find a counterexample, <code>implies</code> lets us abort a proof as soon as we prove that no counterexample exists:</p> <figure><pre><code><span>implies</span> <span>conquer</span> <span>x</span> <span>=</span> <span>conquer</span> <span>divide</span> <span>delta</span> <span>conquer</span> <span>x</span> <span>=</span> <span>x</span> <span>divide</span> <span>delta</span> <span>x</span> <span>conquer</span> <span>=</span> <span>x</span> <span>delta</span> <span>x</span> <span>=</span> <span>(</span><span>x</span><span>,</span><span>x</span><span>)</span></code></pre></figure><p>Both optimizations are used here: The first in <code>findD</code> to abort searching when we find an element, and the second we’ll add explicitly:</p> <figure><pre><code><span>minimizeD</span> <span>::</span> <span>(</span><span>Refinable</span> <span>f</span><span>,</span> <span>Decidable</span> <span>f</span><span>,</span> <span>Reducible</span> <span>a</span><span>)</span> <span>=&gt;</span> <span>f</span> <span>a</span> <span>-&gt;</span> <span>f</span> <span>a</span> <span>minimizeD</span> <span>pred</span> <span>=</span> <span>divide</span> <span>(</span><span>reductions</span> <span>&amp;&amp;&amp;</span> <span>id</span><span>)</span> <span>(</span><span>findD</span> <span>checkChild</span><span>)</span> <span>pred</span> <span>where</span> <span>checkChild</span> <span>=</span> <span>pred</span> <span>`</span><span>implies</span><span>`</span> <span>minimizeD</span> <span>pred</span></code></pre></figure><p>Now the numbers are almost equal:</p> <figure><pre><code><span>PerftTest</span> <span>(</span><span>r1bqkbnr</span><span>/</span><span>pppppppp</span><span>/</span><span>n7</span><span>/</span><span>8</span><span>/</span><span>8</span><span>/</span><span>4</span><span>P3</span><span>/</span><span>PPPP1PPP</span><span>/</span><span>RNBQKBNR</span> <span>w</span> <span>KQkq</span> <span>-</span> <span>0</span> <span>1</span><span>,</span><span>1</span><span>)</span> <span>14</span> <span>*</span><span>Main</span><span>&gt;</span> <span>timeTicks</span> <span>$</span> <span>minimizeD''</span> <span>$</span> <span>testCase</span> <span>3</span> <span>Just</span> <span>(</span><span>PerftTest</span> <span>(</span><span>r1bqkbnr</span><span>/</span><span>pppppppp</span><span>/</span><span>n7</span><span>/</span><span>8</span><span>/</span><span>8</span><span>/</span><span>4</span><span>P3</span><span>/</span><span>PPPP1PPP</span><span>/</span><span>RNBQKBNR</span> <span>w</span> <span>KQkq</span> <span>-</span> <span>0</span> <span>1</span><span>,</span><span>1</span><span>))</span> <span>15</span></code></pre></figure><p>The extra node is because <code>minimizeD</code> scans the root node, while <code>minimize</code> immediately recurses into it. Thus, we’ve inadvertently fixed a minor inconsistency in <code>minimize</code>.</p> <h2 id="ffi">FFI</h2> <p>The last component is the interface between the C++ engine and the Haskell control code. To start with, here are C++ functions I’ll expose:</p> <figure><pre><code><span>#include "chess.cpp" </span> <span>extern</span> <span>"C"</span> <span>{</span> <span>void</span> <span>new_game_w</span><span>(</span><span>gamestate</span> <span>*</span><span>g</span><span>)</span> <span>{</span> <span>*</span><span>g</span> <span>=</span> <span>new_game</span><span>();</span> <span>}</span> <span>void</span> <span>print_move_w</span><span>(</span><span>move</span> <span>*</span><span>m</span><span>,</span> <span>char</span> <span>*</span><span>buffer</span><span>)</span> <span>{</span> <span>print_move</span><span>(</span><span>*</span><span>m</span><span>,</span> <span>buffer</span><span>);</span> <span>}</span> <span>void</span> <span>print_fen_w</span><span>(</span><span>gamestate</span> <span>*</span><span>g</span><span>,</span> <span>char</span> <span>*</span><span>buffer</span><span>)</span> <span>{</span> <span>print_fen</span><span>(</span><span>*</span><span>g</span><span>,</span> <span>buffer</span><span>);</span> <span>}</span> <span>void</span> <span>mkIterator_w</span><span>(</span><span>gamestate</span> <span>*</span><span>g</span><span>,</span> <span>iterator</span> <span>*</span><span>i</span><span>)</span> <span>{</span> <span>*</span><span>i</span> <span>=</span> <span>mkIterator</span><span>(</span><span>*</span><span>g</span><span>);</span> <span>}</span> <span>void</span> <span>advance_iterator_w</span><span>(</span><span>gamestate</span> <span>*</span><span>g</span><span>,</span> <span>iterator</span> <span>*</span><span>i</span><span>,</span> <span>iterator</span> <span>*</span><span>result</span><span>)</span> <span>{</span> <span>*</span><span>result</span> <span>=</span> <span>advance_iterator</span><span>(</span><span>*</span><span>g</span><span>,</span> <span>*</span><span>i</span><span>);</span> <span>}</span> <span>uint64_t</span> <span>perft_w</span><span>(</span><span>gamestate</span> <span>*</span><span>g</span><span>,</span> <span>int</span> <span>depth</span><span>)</span> <span>{</span> <span>return</span> <span>perft</span><span>(</span><span>*</span><span>g</span><span>,</span> <span>depth</span><span>);</span> <span>}</span> <span>void</span> <span>apply_move_w</span><span>(</span><span>gamestate</span> <span>*</span><span>g</span><span>,</span> <span>move</span> <span>*</span><span>m</span><span>,</span> <span>gamestate</span> <span>*</span><span>result</span><span>)</span> <span>{</span> <span>*</span><span>result</span> <span>=</span> <span>apply_move</span><span>(</span><span>*</span><span>g</span><span>,</span> <span>*</span><span>m</span><span>);</span> <span>}</span> <span>int</span> <span>is_iterator_finished_w</span><span>(</span><span>iterator</span> <span>*</span><span>i</span><span>)</span> <span>{</span> <span>return</span> <span>is_iterator_finished</span><span>(</span><span>*</span><span>i</span><span>)</span> <span>?</span> <span>1</span> <span>:</span> <span>0</span><span>;</span> <span>}</span> <span>void</span> <span>dereference_iterator_w</span><span>(</span><span>iterator</span> <span>*</span><span>i</span><span>,</span> <span>move</span> <span>*</span><span>m</span><span>)</span> <span>{</span> <span>*</span><span>m</span> <span>=</span> <span>dereference_iterator</span><span>(</span><span>*</span><span>i</span><span>);</span> <span>}</span> <span>};</span></code></pre></figure><p>Since GHC doesn’t support passing <code>struct</code> values on the stack, I allocate temporary memory and pass in a pointer for the C++ code to write to.</p> <figure><pre><code><span>newtype</span> <span>Gamestate</span> <span>=</span> <span>Gamestate</span> <span>BS</span><span>.</span><span>ByteString</span> <span>newtype</span> <span>Iterator</span> <span>=</span> <span>Iterator</span> <span>BS</span><span>.</span><span>ByteString</span> <span>newtype</span> <span>Move</span> <span>=</span> <span>Move</span> <span>BS</span><span>.</span><span>ByteString</span> <span>instance</span> <span>Show</span> <span>Gamestate</span> <span>where</span> <span>show</span> <span>g</span> <span>=</span> <span>printFen</span> <span>g</span> <span>instance</span> <span>Show</span> <span>Move</span> <span>where</span> <span>show</span> <span>m</span> <span>=</span> <span>printMove</span> <span>m</span> <span>pokeBs</span> <span>::</span> <span>Ptr</span> <span>a</span> <span>-&gt;</span> <span>BS</span><span>.</span><span>ByteString</span> <span>-&gt;</span> <span>IO</span> <span>()</span> <span>pokeBs</span> <span>ptr</span> <span>bs</span> <span>=</span> <span>BS</span><span>.</span><span>useAsCStringLen</span> <span>bs</span> <span>$</span> <span>\</span><span>(</span><span>src</span><span>,</span> <span>len</span><span>)</span> <span>-&gt;</span> <span>copyBytes</span> <span>(</span><span>castPtr</span> <span>ptr</span><span>)</span> <span>src</span> <span>len</span> <span>instance</span> <span>Storable</span> <span>Gamestate</span> <span>where</span> <span>sizeOf</span> <span>_</span> <span>=</span> <span>80</span> <span>alignment</span> <span>_</span> <span>=</span> <span>8</span> <span>peek</span> <span>ptr</span> <span>=</span> <span>Gamestate</span> <span>&lt;$&gt;</span> <span>BS</span><span>.</span><span>packCStringLen</span> <span>(</span><span>castPtr</span> <span>ptr</span><span>,</span> <span>sizeOf</span> <span>(</span><span>undefined</span> <span>::</span> <span>Gamestate</span><span>))</span> <span>poke</span> <span>ptr</span> <span>(</span><span>Gamestate</span> <span>bs</span><span>)</span> <span>=</span> <span>pokeBs</span> <span>ptr</span> <span>bs</span> <span>instance</span> <span>Storable</span> <span>Move</span> <span>where</span> <span>sizeOf</span> <span>_</span> <span>=</span> <span>16</span> <span>alignment</span> <span>_</span> <span>=</span> <span>8</span> <span>peek</span> <span>ptr</span> <span>=</span> <span>Move</span> <span>&lt;$&gt;</span> <span>BS</span><span>.</span><span>packCStringLen</span> <span>(</span><span>castPtr</span> <span>ptr</span><span>,</span> <span>sizeOf</span> <span>(</span><span>undefined</span> <span>::</span> <span>Move</span><span>))</span> <span>poke</span> <span>ptr</span> <span>(</span><span>Move</span> <span>bs</span><span>)</span> <span>=</span> <span>pokeBs</span> <span>ptr</span> <span>bs</span> <span>instance</span> <span>Storable</span> <span>Iterator</span> <span>where</span> <span>sizeOf</span> <span>_</span> <span>=</span> <span>80</span> <span>alignment</span> <span>_</span> <span>=</span> <span>8</span> <span>peek</span> <span>ptr</span> <span>=</span> <span>Iterator</span> <span>&lt;$&gt;</span> <span>BS</span><span>.</span><span>packCStringLen</span> <span>(</span><span>castPtr</span> <span>ptr</span><span>,</span> <span>sizeOf</span> <span>(</span><span>undefined</span> <span>::</span> <span>Iterator</span><span>))</span> <span>poke</span> <span>ptr</span> <span>(</span><span>Iterator</span> <span>bs</span><span>)</span> <span>=</span> <span>pokeBs</span> <span>ptr</span> <span>bs</span></code></pre></figure><p>Each wrapped function follows a regular pattern of allocating memory, writing arguments to it, calling the function, and reading the result. For brevity, I’ll only show a single one:</p> <figure><pre><code><span>applyMove</span> <span>::</span> <span>Gamestate</span> <span>-&gt;</span> <span>Move</span> <span>-&gt;</span> <span>Gamestate</span> <span>applyMove</span> <span>g</span> <span>m</span> <span>=</span> <span>unsafePerformIO</span> <span>$</span> <span>alloca</span> <span>$</span> <span>\</span><span>g_ptr</span> <span>-&gt;</span> <span>alloca</span> <span>$</span> <span>\</span><span>m_ptr</span> <span>-&gt;</span> <span>alloca</span> <span>$</span> <span>\</span><span>g'_ptr</span> <span>-&gt;</span> <span>do</span> <span>poke</span> <span>g_ptr</span> <span>g</span> <span>poke</span> <span>m_ptr</span> <span>m</span> <span>applyMove_w</span> <span>g_ptr</span> <span>m_ptr</span> <span>g'_ptr</span> <span>peek</span> <span>g'_ptr</span></code></pre></figure><p>That covers my own chess engine. We also need to call out to Rocechess to use as a reference. Here’s the function that does that:</p> <figure><pre><code><span>{- Standard output from the "roce38" process looks like: Roce version: 0.0380 - Roman's Own Chess Engine Copyright (C) 2003-2007 Roman Hartmann, Switzerland. All rights reserved. warning: couldn't open Roce.cfg roce: roce: Perft (3): 8902, Time: 0.001 s -}</span> <span>runRoceCommands</span> <span>::</span> <span>[</span><span>String</span><span>]</span> <span>-&gt;</span> <span>([</span><span>String</span><span>]</span> <span>-&gt;</span> <span>a</span><span>)</span> <span>-&gt;</span> <span>IO</span> <span>a</span> <span>runRoceCommands</span> <span>commands</span> <span>parseOutput</span> <span>=</span> <span>do</span> <span>(</span><span>Just</span> <span>hin</span><span>,</span> <span>Just</span> <span>hout</span><span>,</span> <span>_</span><span>,</span> <span>ph</span><span>)</span> <span>&lt;-</span> <span>createProcess</span> <span>$</span> <span>(</span><span>proc</span> <span>"./roce38"</span> <span>[]</span><span>)</span> <span>{</span> <span>std_in</span> <span>=</span> <span>CreatePipe</span><span>,</span> <span>std_out</span> <span>=</span> <span>CreatePipe</span><span>,</span> <span>std_err</span> <span>=</span> <span>Inherit</span> <span>}</span> <span>hSetBuffering</span> <span>hout</span> <span>NoBuffering</span> <span>hSetBuffering</span> <span>hin</span> <span>NoBuffering</span> <span>forM_</span> <span>commands</span> <span>$</span> <span>\</span><span>command</span> <span>-&gt;</span> <span>do</span> <span>hPutStr</span> <span>hin</span> <span>command</span> <span>hPutChar</span> <span>hin</span> <span>'</span><span>\n</span><span>'</span><span> </span><span> </span><span> </span> <span>output</span> <span>&lt;-</span> <span>hGetContents</span> <span>hout</span> <span>tickCounter</span> <span>return</span> <span>$</span> <span>parseOutput</span> <span>$</span> <span>drop</span> <span>6</span> <span>$</span> <span>lines</span> <span>output</span> <span>reference_perft_w</span> <span>::</span> <span>Gamestate</span> <span>-&gt;</span> <span>Int32</span> <span>-&gt;</span> <span>IO</span> <span>Word64</span> <span>reference_perft_w</span> <span>g</span> <span>d</span> <span>=</span> <span>let</span> <span>commands</span> <span>=</span> <span>[</span> <span>"setboard "</span> <span>++</span> <span>printFen</span> <span>g</span><span>,</span> <span>"perft "</span> <span>++</span> <span>show</span> <span>d</span><span>,</span> <span>"quit"</span> <span>]</span> <span>parseOutput</span> <span>(</span><span>perft_line</span> <span>:</span> <span>_</span><span>)</span> <span>=</span> <span>let</span> <span>perft_word</span> <span>=</span> <span>splitOneOf</span> <span>" ,"</span> <span>perft_line</span> <span>!!</span> <span>2</span> <span>perft</span> <span>=</span> <span>read</span> <span>perft_word</span> <span>in</span> <span>perft</span> <span>in</span> <span>runRoceCommands</span> <span>commands</span> <span>parseOutput</span> <span>reference_perft</span> <span>g</span> <span>d</span> <span>=</span> <span>unsafePerformIO</span> <span>$</span> <span>reference_perft_w</span> <span>g</span> <span>d</span></code></pre></figure><p>Note the use of <code>tickCounter</code> to measure the number of times we invoke this expensive command.</p> <h2 id="why-not-arbitrary">Why not Arbitrary?</h2> <p>I mainly used <code>Reducible</code> in this article because it is a simpler typeclass than <code>Arbitrary</code>. There’s not much reason to use <code>Reducible</code> over <code>Arbitrary</code> for our chess example. This is because <code>perft</code> test failures are <strong>monotone</strong>: A successful test implies that all transitive children are successful. A real example that isn’t monotone that I’ve come across would be a compiler that emitted a jump to a non-existent label:</p> <figure><pre><code><span>label1:</span> <span>goto</span> <span>label2</span> <span>goto</span> <span>label1</span> </code></pre></figure><p>In this case, we want <code>minimize</code> to return just <code>goto label2</code>, because that is the smallest statement that exhibits the bug. However, <code>shrinkList</code> in the <code>QuickCheck</code> package employs an optimization that drops an exponentially-decreasing percentage of elements from a list for performance, which means it could drop <code>label1</code> and <code>goto label2</code>:</p> <figure><pre><code><span>goto</span> <span>label1</span></code></pre></figure><p>This still has a jump to a non-existent label, but only because our shrinking dropped it. It’s possible to restore the monotone property in full by defining the error condition in sufficient detail; but it can also be restored by only reducing to immediate children. It may be easier to check that <code>label1</code> can be excluded from the immediate children because <code>goto label1</code> depends on it, than to check whether an entire set of statements can be safely removed.</p> <h2 id="conclusion">Conclusion</h2> <p>We generalized a function for reducing a failing testcase to a minimal state. In particular, there is an instance for <code>Compose</code> allowing the resulting algorithm to work over any <code>Monad</code> or <code>Applicative</code>. This allows one to augment the reduction process with benchmarking information, experiment with clearly-defined optimizations, and test novel reduction algorithms on simple pure values which automatically generalize to the full application.</p> <p>The code for this article is available on <a href="https://github.com/MichaelBurge/delta-debugging-example">Github</a>.</p> <p>Future articles may cover:</p> <ul>
<li>Further optimizing testcase reduction tools</li> <li>Generalizing other common utilities with abstract typeclasses</li> <li>More techniques for debugging programs</li> </ul>
</div> </div><br>
<br>
<br>
<br>
<br>
<br>
via www.michaelburge.us http://bit.ly/2WgYoTx<br>
<br>
June 6, 2019 at 02:48PM
